Refactored Field Validation.

BeksOmega opened the pull request:
<!--
  - Thanks for submitting code to Blockly!  Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## The basics

<!-- TODO: Verify the following, checking each box with an 'x' between the brackets: [x] -->

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

## The details
### Resolves

<!-- TODO: What Github issue does this resolve? Please include a link. -->
#2169 

### Proposed Changes

<!-- TODO: Describe what this Pull Request does.  Include screenshots if applicable. -->
Changes field validation to work like discussed [here](https://github.com/google/blockly/issues/2169#issuecomment-487407878).

### Reason for Changes

<!--TODO: Explain why these changes should be made.  Include screenshots if applicable. -->
ATM fields only expect one value, and they expect it to be text, this is not true of all fields.
Also field validation is pretty all over the place, this should unify it to make it easier to debug.

### Test Coverage

<!-- TODO: Please show how you have added tests to cover your changes,
  -        or tell us how you tested it. For each systems you tested,
  -        uncomment the systems in the list below.
  -->

For now N/A

Tested on:
<!-- * Desktop Chrome -->
<!-- * Desktop Firefox -->
<!-- * Desktop Safari -->
<!-- * Desktop Opera -->
<!-- * Windows Internet Explorer 10 -->
<!-- * Windows Internet Explorer 11 -->
<!-- * Windows Edge -->

<!--
* Smartphone/Tablet/Chromebook (please complete the following information):
  * Device: [e.g. iPhone6]
  * OS: [e.g. iOS8.1]
  * Browser [e.g. stock browser, safari]
  * Version [e.g. 22]
-->

### Additional Information

This isn't finished yet, but it's going to be quite a whopper of a PR, so I wanted to put it up so it could be reviewed in stages.

BeksOmega commented:
The first leg!

### Resolves
#2435

### Proposed Changes
Refactored FieldTextInput to conform to the new structure.

### Test Coverage

Added text validator test blocks to the playground.

* The "null validator" block always returns null (invalid) for all inputs
* The "remove 'a' validator" block removes any 'a' characters from the input value.
* The "'b' -> null validator" block returns null (invalid) if it detects a 'b' character in the input.

![Validation_TextInput_Null](https://user-images.githubusercontent.com/25440652/57185203-02979780-6e7c-11e9-9d24-4813aec7064e.gif)
![Validation_TextInput_A](https://user-images.githubusercontent.com/25440652/57185206-09bea580-6e7c-11e9-8614-d90dcca81167.gif)
![Validation_TextInput_B](https://user-images.githubusercontent.com/25440652/57185207-0c20ff80-6e7c-11e9-8bfd-8fb7491b4646.gif)

(To do the logging change the logger function on the playground to look like the below, and enable logging events)
```
function logger(e) {
  if (e.type == Blockly.Events.BLOCK_CHANGE && e.element == 'field') {
    console.log('new value: ', e.newValue);
  }
}
```

Tested setting value programmatically:
* Value = "ab"
![Validator_TextInput_ab](https://user-images.githubusercontent.com/25440652/57185273-0d9ef780-6e7d-11e9-95e9-0c91342d858a.jpg)

* Value = "a"
![Validator_TextInput_a](https://user-images.githubusercontent.com/25440652/57185272-0677e980-6e7d-11e9-90ab-aae44d370656.jpg)

* Value = "b"
![Validator_TextInput_b](https://user-images.githubusercontent.com/25440652/57185274-11327e80-6e7d-11e9-87f9-81fcb97601cc.jpg)

Tested undo/redo and events (see #1508):
![Validation_TextInput_Undo](https://user-images.githubusercontent.com/25440652/57185316-c5cca000-6e7d-11e9-953e-e8dde5041521.gif)

There is one ctrl-z for each change in the workspace.

### Additional Information

I know all of the new text input stuff works, but I'm not sure if it's readable enough. Please tell me if there are places that are confusing!

Also @rachel-fenichel [here ](https://github.com/google/blockly/issues/2169#issuecomment-488471207)you mentioned

> tests that are the same before and after, to verify that existing implementations continue to work.

Do you mean like a test-field that overrides setValue or setText, and checks if it still gets called correctly? Or did you have something else you were thinking of?

BeksOmega commented:
Second leg!

### Proposed changes

Refactored FieldAngle to conform to the new structure.

### Test Coverage

Added angle validator blocks to the playground.

* The "always null" block always returns null (invalid) for all inputs.
* The "force mult of 30" block returns the closest multiple of 30 to the input.
* The "not mult of 30 -> null" block returns null (invalid) if the input is not a multiple of thirty.
![Validation_Angle_Null](https://user-images.githubusercontent.com/25440652/57199749-6763f800-6f37-11e9-9963-dfc8e2d73ed9.gif)
![Validation_Angle_Force30](https://user-images.githubusercontent.com/25440652/57199752-69c65200-6f37-11e9-9fc1-472a82384dd9.gif)
![Validation_Angle_Null30](https://user-images.githubusercontent.com/25440652/57199753-6b901580-6f37-11e9-9d7a-79ae2fa104d2.gif)

Tested setting value programmatically:
* Value = 15:
![Validator_Angle_15](https://user-images.githubusercontent.com/25440652/57199795-1a345600-6f38-11e9-9681-a7c006765b86.jpg)

### Additional Information

FieldAngle's value_ property now contains a number instead of text. This should act the same as before with regard to generators. It should also act the same with regard to validators because devs should already be converting the text into a number inside their validator. But now future peoples won't have to do that!

BeksOmega commented:
Third leg!

### Proposed Changes

Refactored FieldCheckbox to conform to the new structure.

### Test Coverage

Added checkbox validator blocks to the playground.

* The "always null" block always returns null (invalid) for all inputs.
* The "force match" block has a validator on the end-most checkbox, which makes it match the start-most checkbox.
* The "not match -> null" block  has a validator on the end-most checkbox, which returns null (invalid) if the input does not match the start-most checkbox.
![Validation_Checkbox_Null](https://user-images.githubusercontent.com/25440652/57249384-a3f82800-6ff9-11e9-885d-c5d83197c40a.gif)
![Validation_Checkbox_Match](https://user-images.githubusercontent.com/25440652/57249390-a5c1eb80-6ff9-11e9-944b-ba8d9b1ff2e6.gif)
![Validation_Checkbox_MatchNull](https://user-images.githubusercontent.com/25440652/57249394-a8244580-6ff9-11e9-9405-87015ca7da9f.gif)

Tested setting value programatically:
* Original State was True then False
![Validator_Checkbox_TF_Before](https://user-images.githubusercontent.com/25440652/57250132-9479de80-6ffb-11e9-8e67-17375adb843e.jpg)

* Set to TRUE
![Validator_Checkbox_TF_True](https://user-images.githubusercontent.com/25440652/57250141-993e9280-6ffb-11e9-8951-0bf9ddfd0e9f.jpg)

* Set to FALSE
![Validator_Checkbox_TF_False](https://user-images.githubusercontent.com/25440652/57250147-9ba0ec80-6ffb-11e9-8c1a-629fb64cf24a.jpg)

* Original State was False then True
![Validator_Checkbox_FT_Before](https://user-images.githubusercontent.com/25440652/57250159-9f347380-6ffb-11e9-9baf-471cd953e4ef.jpg)

* Set to True
![Validator_Checkbox_FT_True](https://user-images.githubusercontent.com/25440652/57250168-a196cd80-6ffb-11e9-97bf-969fc7c1ee77.jpg)

* Set to False
![Validator_Checkbox_FT_False](https://user-images.githubusercontent.com/25440652/57250170-a3f92780-6ffb-11e9-8457-2a4c4e07cd2d.jpg)

### Additional Information

It may be worth considering deprecating the current checkbox in favor of one that contains an actual true/false value, probably not though.

BeksOmega commented:
Colour fields.

###Proposed Changes
Refactored FieldCheckbox to conform to the new structure.

###Test Coverage
Added checkbox validator blocks to the playground.

* The "always null" block always returns null (invalid) for all inputs.
* The "force red" block changes the value to always have a full red value.
* The "not red -> null" block returns null (invalid) if the colour does not have a full red value.
![Validation_Colour_Null](https://user-images.githubusercontent.com/25440652/57254047-2508ec80-7005-11e9-951b-2a26a4e996e7.gif)
![Validation_Colour_ForceRed](https://user-images.githubusercontent.com/25440652/57254056-29cda080-7005-11e9-9b11-6da2fb5882c5.gif)
![Validation_Colour_RedNull](https://user-images.githubusercontent.com/25440652/57254058-2b976400-7005-11e9-9303-f3f16c76a737.gif)

Tested setting value programmatically:
* Set Value #ffffff
![Validator_Color_FFFFFF](https://user-images.githubusercontent.com/25440652/57254342-e32c7600-7005-11e9-99d3-216ce15f7b37.jpg)

* Set Value #00ffff
![Validator_Color_00FFFF](https://user-images.githubusercontent.com/25440652/57254378-f7707300-7005-11e9-9025-dc6499a1e69c.jpg)

### Additional Information
N/A

BeksOmega commented:
Date fields.

### Proposed Changes
Refactored FieldDate to conform to the new structure.

### Test Coverage
Added date validator blocks to the playground.

* The "always null" block always returns null (invalid) for all inputs.
* The "force day 20s" block changes the value to always have a day in the 20s.
* The "not 20s-> null" block returns null (invalid) if the date's day value is not in the 20s.
![Validation_Date_Null](https://user-images.githubusercontent.com/25440652/57258995-f1cd5a00-7012-11e9-92d1-7937f12ea460.gif)
![Validation_Date_Force20s](https://user-images.githubusercontent.com/25440652/57259000-f4c84a80-7012-11e9-9d4d-e2a792926ade.gif)
![Validation_Date_20sNull](https://user-images.githubusercontent.com/25440652/57259002-f6920e00-7012-11e9-8600-d5e714a1f65b.gif)

Tested setting the value programmatically:
Set value to '2020-02-15':
![Validator_Date_2020-02-15](https://user-images.githubusercontent.com/25440652/57259116-55f01e00-7013-11e9-90ea-511e773c36bd.jpg)

BeksOmega commented:
Date fields.

### Proposed Changes
Refactored FieldDate to conform to the new structure.

### Test Coverage
Added date validator blocks to the playground.

* The "always null" block always returns null (invalid) for all inputs.
* The "force day 20s" block changes the value to always have a day in the 20s.
* The "not 20s-> null" block returns null (invalid) if the date's day value is not in the 20s.
![Validation_Date_Null](https://user-images.githubusercontent.com/25440652/57258995-f1cd5a00-7012-11e9-92d1-7937f12ea460.gif)
![Validation_Date_Force20s](https://user-images.githubusercontent.com/25440652/57259000-f4c84a80-7012-11e9-9d4d-e2a792926ade.gif)
![Validation_Date_20sNull](https://user-images.githubusercontent.com/25440652/57259002-f6920e00-7012-11e9-8600-d5e714a1f65b.gif)

Tested setting the value programmatically:
Set value to '2020-02-15':
![Validator_Date_2020-02-15](https://user-images.githubusercontent.com/25440652/57259116-55f01e00-7013-11e9-90ea-511e773c36bd.jpg)

BeksOmega commented:
Dropdown fields.

### Proposed Changes
Refactored FieldDropdown to conform to the new structure.

### Test Coverage
Added dropdown validator blocks to the playground.

* The "always null" block always returns null (invalid) for all inputs.
* The "force1s" block changes the language-neutral value to always start with a one.
* The "not 1s-> null" block returns null (invalid) if the value does not start with a one.
![Validation_Dropdown_Null](https://user-images.githubusercontent.com/25440652/57306345-811c5100-7097-11e9-857e-d6c007cb00e1.gif)
![Validation_Dropdown_Force1s](https://user-images.githubusercontent.com/25440652/57306350-837eab00-7097-11e9-95e3-b8453210f096.gif)
![Validation_Dropdown_1sNull](https://user-images.githubusercontent.com/25440652/57306357-84afd800-7097-11e9-80b1-ea418a7378ac.gif)

Tested setting value programmaticly:
* Value 1B
![Validator_Dropdown_1B](https://user-images.githubusercontent.com/25440652/57306417-a315d380-7097-11e9-854d-a967f760c417.jpg)

* Value 2B
![Validator_Dropdown_2B](https://user-images.githubusercontent.com/25440652/57306477-bde84800-7097-11e9-8e8b-efb0c68d9035.jpg)

* Value 3B
![Validator_Dropdown_3B](https://user-images.githubusercontent.com/25440652/57306479-c0e33880-7097-11e9-8263-d313b96bb450.jpg)
Correctly got log message:
`Cannot set the dropdown's value to an unavailable option. Block type: test_validators_dropdown_null, Field name: INPUT`

rachel-fenichel commented:
Thanks for splitting this up by field type.  It's a lot of code!

This stuff looks like it should be testable in Mocha--you would be able to set up a field, call setValue with a variety of values, and then validate the resulting value and text.  Since this should be a no-op in terms of existing field behaviour, you would be able to run the tests before and after the change and get the same results.

BeksOmega commented:
> This stuff looks like it should be testable in Mocha--you would be able to set up a field, call setValue with a variety of values, and then validate the resulting value and text. Since this should be a no-op in terms of existing field behaviour, you would be able to run the tests before and after the change and get the same results.

Sweet! I think set up tests in a separate PR and then rebase on top of it before this gets merged, just to keep things a little more organized.

BeksOmega commented:
Image fields.

### Proposed Changes
Refactored FieldImage to conform to the new structure.

### Test Coverage
Note: Image fields don't take in validators, so we don't need to test those.

I added a default image field block to the playground, with a button to change the image src & alt text. I also added a "toggle collapsed" button so we can see how the blocks look when collapsed.
![Validation_Image](https://user-images.githubusercontent.com/25440652/57328576-a1afcf80-70c6-11e9-9870-57af2c2e807e.gif)

### Additional Information

There's an issue with collapsing because setVisible is changing the field's private size_.width property to 0 when really that should just be handled inside getSize. So some fields like colour, image, etc have to override render to reset the field size. I'll come back and clean that up in a different PR.

Also the checkbox isn't handling collapsing properly, I'll have to check on that.

rachel-fenichel commented on core/field.js:
```diff
@@ -633,39 +638,114 @@ Blockly.Field.prototype.forceRerender = function() {
+ * @type {*}
+ * @protected
+ */
+Blockly.Field.prototype.value_ = null;
```
Move this to the top of the file so it's not lost in here with the functions.

BeksOmega replied:
Fixed

rachel-fenichel commented on core/field.js:
```diff
@@ -633,39 +638,114 @@ Blockly.Field.prototype.forceRerender = function() {
+ * Used to validate a value. Returns input by default. Can be overridden by
+ * subclasses, see FieldDropdown.
+ * @param {*} newValue The value to be validated.
+ * @return {*} The validated value, same as input by default.
```
As I recall, the possible return values are null, undefined, or the validated value, where null means it's invalid and undefined means...something.  Is that accurate?

Either way, please update the comment to say what null means.

BeksOmega replied:
Null means its invalid, which by default means the field reverts to its previous value.

Undefined doesn't mean anything at the moment, but most fields will revert to a default value (default meaning the value they would have if they had been constructed with no arguements, not any value they /were/ constructed with) when passed a value they can't parse (such as undefined, NaN, 'bad', etc).

This seems like a bug, since not all fields work this way. What I want to go back through and do is have fields' class validators validate anything they can't parse (undefined, NaN, 'bad', etc) to null (invalid). Any default values would be handled in the constructor.

But that's a bit besides the point, if we want to add more special cases besides 'invalid' I would prefer to move to using constants rather than keywords, so it's more extendable in the future.

rachel-fenichel commented on core/field.js:
```diff
@@ -441,15 +446,24 @@ Blockly.Field.prototype.updateColour = function() {
+  try {
+    this.textElement_.childNodes[0].nodeValue = this.getDisplayText_();
+  } catch (error) {
+    if (this.textElement_.childNodes.length == 0) {
```
Why do this as a try-catch instead of checking the length of `childNodes` before trying to make the change?

BeksOmega replied:
So this happened because of how the angle field's degree symbol is static rather than reappended every time render is called. It used to be reappended everytime because when the textElement_'s textContent was set it would get kicked. So instead of setting the textContent it just sets the nodeValue of the textElement_'s first text node. Buuut if the text element doesn't have a first text node everything breaks. This can happen if a developer writes their initView function incorrectly.

TL:DR A series of changes meant I had to add this so developers know if they derp something up.

Really this needs a better solution though, how would you feel about some sort of textGroup_ that the textElement_ as well as things like degree symbols could be children of? Then render could go back to how it was before, and this could be removed.

BeksOmega replied:
Opinions @rachel-fenichel ?

rachel-fenichel replied:
Either that, or somehow move the special sauce back into field angle (even if it means putting it back to a dynamic state).  I'd like to avoid having a throw like this in the general field render path.

BeksOmega replied:
I ended up reverting it because I changing it to have a new DOM structure would mean reworking some CSS, and I didn't want to do that in this PR.

I created an [issue ](https://github.com/google/blockly/issues/2484)to track fixing this (because it bugs me), but if you don't think it's a big deal I'm fine with closing it.

rachel-fenichel commented on core/field.js:
```diff
@@ -633,39 +648,109 @@ Blockly.Field.prototype.forceRerender = function() {
-  if (validated !== null) {
-    newValue = validated;
+
+  var userValidator = this.getValidator();
```
Nit: maybe "localValidator" or "fieldValidator".  "userValidator" feels misleading.

BeksOmega replied:
Fixed, I picked localValidator.

rachel-fenichel commented on core/field_angle.js:
```diff
@@ -35,21 +35,17 @@ goog.require('goog.userAgent');
  * @constructor
  */
 Blockly.FieldAngle = function(opt_value, opt_validator) {
-  // Add degree symbol: '360°' (LTR) or '°360' (RTL)
```
Making sure I understand:
Removing the degree symbol stuff here isn't specifically because of the change to the validator--it's because it was previously dynamic (depended on RTL) and is now always on the same side of the number, so it can just be an additional textNode that we don't change.  

Is that right?

BeksOmega replied:
Correct! I had to switch up the rendering because of how render_ changed to handle htmlInput being red or not.

rachel-fenichel replied:
Note to self: I still need to dig into the rendering change on textinput, and how it reflects back into other fields.

rachel-fenichel commented on core/field_angle.js:
```diff
@@ -121,18 +117,23 @@ Blockly.FieldAngle.WRAP = 360;
+ */
+Blockly.FieldAngle.prototype.init = function() {
+  Blockly.FieldAngle.superClass_.init.call(this);
+  var symbol = Blockly.utils.createSvgElement('tspan', {}, null);
```
Add a comment: // Add the degree symbol to the left of the number, even in RTL (issue #2380)

BeksOmega replied:
Fixed

rachel-fenichel commented on core/field_angle.js:
```diff
@@ -344,7 +335,14 @@ Blockly.FieldAngle.prototype.classValidator = function(text) {
+  Blockly.FieldAngle.superClass_.doValueUpdate_.call(this, newValue);
+  // Always render the input angle.
+  this.text_ = String(newValue);
+  this.forceRerender();
```
You call `forceRerender` in `Field.setValue`, just after `doValueUpdate_`.  Why the duplication?

BeksOmega replied:
Not sure why this ended up down here, it should be inside the onMouseMove function. 

Fixed.

rachel-fenichel commented on core/field_checkbox.js:
```diff
@@ -92,49 +93,72 @@ Blockly.FieldCheckbox.prototype.init = function() {
+  this.value_ = this.convertValueToBool_(newValue);
+  // Update visual.
+  if (this.checkElement_) {
+    this.checkElement_.style.display = this.value_ ? 'block' : 'none';
```
This UI update looks like it should happen in the rerender, with `doValueUpdate` just storing the new value and marking the field as dirty.

rachel-fenichel replied:
Actually, I'm seeing this as a pattern across other fields as well.  More work is happening in doValueUpdate than I expect, and less in render, which often just sets the width.

BeksOmega replied:
Doing stuff inside doValueUpdate_ or doValueInvalid is actually a little bit of an optimization. Setting isDirty to true (and doing work inside render_) means the whole block has to rerender, which is unecessary if the fields size doesn't change.

The reason for setting the width inside render is because setVisible is broken at the moment. It's setting the field's size_.width to zero when that should really be handled inside getSize.

rachel-fenichel replied:
Interesting.  Sounds like that's a separate issue, but one worth looking into (avoiding a rerender if the field's size didn't change).  Can you create an issue for it and note which fields will likely need investigation?

BeksOmega replied:
Can do, do you want everything moved to render then @rachel-fenichel ?

rachel-fenichel replied:
Leave it here, and then move it in a followup.

BeksOmega replied:
Sorry, I think I got myself a bit confused, and I'm unsure of exactly what you're looking for.

I thought I was supposed to remove the optimization, and then put up an issue to re-add the optimization once I could prove it was actually more efficient. So I asked:
> do you want everything moved to render then?  

To make sure I was understanding correctly.

But now it seems like I'm not supposed to remove the optimization? Am I'm still supposed to create an issue to investigate the optimization?

I'm cool with doing whatever! I just wanted to make sure I'm not creating any confusing issues. Sorry I had to bug you again, and thank you for the helps!

rachel-fenichel replied:
I think you've made a good case that it's more efficient, and my issue with it is just that the code reads in a confusing way--the meat of rendering the field doesn't always happen in render or a helper called by render.  So I meant 
- you should leave the code as it is right now (correct and efficient).
- we should (separately) add a code path that rerenders only the field and not the block if the field's size doesn't change.
- once that path exists, we should move this code (and similar for other fields) back into the render function or its helpers

Does that make sense?

BeksOmega replied:
> * you should leave the code as it is right now (correct and efficient).
> * we should (separately) add a code path that rerenders only the field and not the block if the field's size doesn't change.
> * once that path exists, we should move this code (and similar for other fields) back into the render function or its helpers
> 
> Does that make sense?

Makes sense! I put up #2490 with both proposed solutions.

rachel-fenichel commented on core/field_checkbox.js:
```diff
@@ -92,49 +93,72 @@ Blockly.FieldCheckbox.prototype.init = function() {
+ * Get the value of this field, either 'TRUE' or 'FALSE'.
+ * @return {string} The value of this field.
+ */
+Blockly.FieldCheckbox.prototype.getValue = function() {
```
While you're here, add a `getValueBoolean` method on `FieldCheckbox` that just returns the actual value, so that developers don't *have* to go back and forth checking string equality.

BeksOmega replied:
Fixed

rachel-fenichel commented on core/field_dropdown.js:
```diff
@@ -396,40 +389,49 @@ Blockly.FieldDropdown.prototype.getValue = function() {
-        this.sourceBlock_, 'field', this.name, this.value_, newValue));
+  if (!isValueValid) {
+    console.warn('Cannot set the dropdown\'s value to an unavailable option.' +
+      ' Block type: ' + this.sourceBlock_.type + ', Field name: ' + this.name);
```
Add the value to the warning message.

BeksOmega replied:
Fixed

rachel-fenichel commented on core/field_textinput.js:
```diff
@@ -234,16 +260,14 @@ Blockly.FieldTextInput.prototype.bindEvents_ = function(htmlInput) {
       Blockly.bindEventWithChecks_(
           htmlInput, 'keypress', this, this.onHtmlInputChange_);
-  htmlInput.onWorkspaceChangeWrapper_ = this.resizeEditor_.bind(this);
-  this.workspace_.addChangeListener(htmlInput.onWorkspaceChangeWrapper_);
```
Why did you remove this?  I think you need it to make sure that the inline editor floats to the right place if you resize the window or workspace while you're editing this text.

BeksOmega replied:
If that's what it was supposed to do it doesn't do that, the workspace doesn't support resize/move/zoom events yet, plus window resize calls hideChaff. As far as I can tell it's just code accretion.

I'll revert it and create an issue to further investigate.

rachel-fenichel commented on core/field_textinput.js:
```diff
@@ -234,16 +260,14 @@ Blockly.FieldTextInput.prototype.bindEvents_ = function(htmlInput) {
   htmlInput.onKeyDownWrapper_ =
       Blockly.bindEventWithChecks_(
           htmlInput, 'keydown', this, this.onHtmlInputKeyDown_);
-  // Bind to keyup -- trap Enter; resize after every keystroke.
```
What about Enter?

BeksOmega replied:
As far as I can tell onHtmlInputChange_ doesn't do anything with enter. I think this was just a typo so I removed it.

rachel-fenichel commented on core/field_textinput.js:
```diff
@@ -371,12 +366,30 @@ Blockly.FieldTextInput.prototype.widgetDispose_ = function() {
+    // Otherwise don't rerender.
+
+    // Call onFinishEditing
+    // TODO: Get rid of this or make it less of a hack.
```
FWIW the case where this is really important is when editing the name of a parameter in a function.  That has code to create a variable with the given name at every keypress, and then it deletes intermediate variables when the user is done editing.

rachel-fenichel commented:
I just added more comments, and I've covered all of the files once so far.
Overall it looks good.  I'd like to straighten out when rendering actually happens, as I mentioned in a few comments.

BeksOmega commented:
Number Fields!

### Proposed changes
Refactored FieldNumber to conform to the new structure.

### Test Coverage
Added number validator blocks to the playground.

* The "always null" block always returns null (invalid) for all inputs.
* The "force mult of 10" block returns the closest multiple of 30 to the input.
* The "not mult of 10-> null" block returns null (invalid) if the input is not a multiple of ten.
![Validation_Number_Null](https://user-images.githubusercontent.com/25440652/57573164-72f06c80-73d8-11e9-83d5-4d4f8c7304cf.gif)
![Validation_Number_Force10](https://user-images.githubusercontent.com/25440652/57573165-74ba3000-73d8-11e9-9db2-e97abea90e7a.gif)
![Validation_Number_10Null](https://user-images.githubusercontent.com/25440652/57573166-75eb5d00-73d8-11e9-937b-35202125e059.gif)

Tested setting value programmatically
* Value = 10
![Validator_Number_10](https://user-images.githubusercontent.com/25440652/57573174-8dc2e100-73d8-11e9-8a0f-e3279ac358a1.jpg)

* Value = 97
![Validator_Number_97](https://user-images.githubusercontent.com/25440652/57573176-90253b00-73d8-11e9-97c9-3d243cba9327.jpg)

Also the precision is still working:
![Validator_Number_Precision](https://user-images.githubusercontent.com/25440652/57573224-3113f600-73d9-11e9-9d79-425d3d97933f.jpg)

### Additional Information

Field number now contains a number value instead of text.

The label fields don't override setValue so those don't need to be refactored. This means I'm just down to variable fields. So my remaining tasks for this PR are:

- [ ]  Refactor Variable Fields
- [ ]  Fix Review Comments
- [ ]  Rebase onto Tests
- [ ]  Pass Tests (Eventually)
- [ ]  Clean up Logging, Formatting, etc
- [ ]  Celebrate!

BeksOmega commented:
Variable fields.

### Proposed Changes
Refactored FieldVariable to conform to the new structure.

### Test Coverage
Added variablevalidator blocks to the playground.

* The "always null" block always returns null (invalid) for all inputs.
* The "force1s" block changes the variable ID to always start with a one.
* The "not 1s-> null" block returns null (invalid) if the variable ID does not start with a one
![Validation_Variable_Null](https://user-images.githubusercontent.com/25440652/57583661-68d97700-7487-11e9-9748-1513e23c94b1.gif)
![Validation_Variable_Force1s](https://user-images.githubusercontent.com/25440652/57583663-6b3bd100-7487-11e9-8657-735ab0e0fdf1.gif)
![Validation_Variable_1sNull](https://user-images.githubusercontent.com/25440652/57583665-6d9e2b00-7487-11e9-8ff9-840a2f9ef517.gif)

Tested setting value programmaticly:
* Value 1B
![Validator_Variable_1B](https://user-images.githubusercontent.com/25440652/57583673-90c8da80-7487-11e9-961b-b2ff72b4fb6e.jpg)

* Value 2B
![Validator_Variable_2B](https://user-images.githubusercontent.com/25440652/57583674-93c3cb00-7487-11e9-9256-de191da14f27.jpg)

* Value 3B
![Validator_Variable_3B](https://user-images.githubusercontent.com/25440652/57583675-96bebb80-7487-11e9-891a-934a83f9cb95.jpg)
& Throws a warning.

BeksOmega commented:
Besides the generator tests failing I think this should be good for another look @rachel-fenichel.

I'm having some trouble figuring out why the generator tests are failing, since they won't deserialize completely inside firefox, but I might take another look at it later.

rachel-fenichel commented on core/field_checkbox.js:
```diff
@@ -91,49 +94,98 @@ Blockly.FieldCheckbox.prototype.initView = function() {
-Blockly.FieldCheckbox.prototype.getValue = function() {
-  return String(this.state_).toUpperCase();
+Blockly.FieldCheckbox.prototype.render_ = function() {
+  this.size_.width = 5;
```
Make a constant (Blockly.FieldCheckbox.WIDTH) to use here and in the constructor.

BeksOmega replied:
Fixed

rachel-fenichel commented on core/field_dropdown.js:
```diff
@@ -386,48 +371,58 @@ Blockly.FieldDropdown.prototype.getOptions = function() {
 /**
- * Get the language-neutral value from this dropdown menu.
- * @return {string} Current text.
+ * Check that the new value is an available option.
```
Remove extra comment.

BeksOmega replied:
Fixed

rachel-fenichel commented on core/field_image.js:
```diff
@@ -166,28 +167,28 @@ Blockly.FieldImage.prototype.setTooltip = function(newTip) {
- * Get the source URL of this image.
- * @return {string} Current text.
- * @override
+ * Ensure that the input value is a string.
```
Ensure that the input value _(the source URL)_ is a string.

BeksOmega replied:
Fixed

rachel-fenichel commented on core/field_number.js:
```diff
@@ -31,22 +31,23 @@ goog.require('Blockly.FieldTextInput');
-  opt_value = (opt_value && !isNaN(opt_value)) ? String(opt_value) : '0';
+  opt_value = this.doClassValidation_(opt_value);
+  if (opt_value === null) {
+    opt_value = 0;
```
What if 0 is outside of the constraints?  Maybe set it to the nearest constraint (min if 0 is below min, max if 0 is above max).

@RoboErikG any preference?

rachel-fenichel replied:
Or do we just say eh, creating it with an invalid initial value is a bad idea? I think there's a similar question for all fields--what if the default value for a text field is the empty string but the validator requires it to be nonempty, for instance?  Is there even a reasonable response?

BeksOmega replied:
I'll wait for @RoboErikG 's response before I change things.

BeksOmega commented on core/field_variable.js:
```diff
@@ -96,28 +104,28 @@ Blockly.FieldVariable.prototype.initModel = function() {
-  } finally {
-    Blockly.Events.enable();
-  }
+  this.setValue(variable.getId());
```
Affirmative

BeksOmega commented on core/field_variable.js:
```diff
@@ -192,30 +207,52 @@ Blockly.FieldVariable.prototype.getVariable = function() {
+
+/**
+ * Update the value of this variable field, as well as its variable and text.
+ * @param {string} newId The id of the new variable.
```
The class validator makes sure it's valid, but then any local validator could make it invalid again. This is a problem with all fields. The setValue template method could be changed so that the class validator runs twice if there is a local validator, but there are performance costs.

rachel-fenichel replied:
Add a comment that it *should* be valid, and that if validators return bad IDs it could break.

BeksOmega replied:
Fixed

BeksOmega commented on core/field_variable.js:
```diff
@@ -248,9 +285,8 @@ Blockly.FieldVariable.prototype.getVariableTypes_ = function() {
-      return workspace.getVariableTypes();
+    if (this.workspace_) {
+      return this.workspace_.getVariableTypes();
     }
```
I don't think so. If the field's variableTypes property isn't defined, all variable types are valid, correct? Based on the logic of typeIsAllowed_, getVariableTypes should just return something falsy (like null) if variableTypes isn't defined.

It might be better to investigate this separately, if you agree I can put up an issue for it.

rachel-fenichel replied:
Confirmed, separate issue but should investigate before the next release.

BeksOmega replied:
Added Issue #2483

BeksOmega commented on core/field.js:
```diff
@@ -644,39 +669,106 @@ Blockly.Field.prototype.forceRerender = function() {
+/**
+ * Used to notify the field an invalid value was input. Can be overiden by
+ * subclasses, see FieldTextInput.
+ * No-op by default.
```
**With regard to the value:**
The field reverts to the value it had before the start of the "edit". For text fields the edit begins when the user opens the input, and ends when the user closes the input. For a colour field each selection is an individual edit.

This means that [this ](https://user-images.githubusercontent.com/25440652/57254058-2b976400-7005-11e9-9303-f3f16c76a737.gif)colour field /doesn't/ revert to its editor-opened value (red):

But [this ](https://user-images.githubusercontent.com/25440652/57185207-0c20ff80-6e7c-11e9-8bfd-8fb7491b4646.gif)text field /does/ revert to its editor-opened value (default) instead of reverting to a value of aaa.

I didn't realize this was inconsistent until you mentioned it, so this may or may not be what you want. Should be pretty easy to change though =)

**With regard to the text:**
Restated in a more general way, I believe you're asking if by default the display will represent the input value, even if it is invalid? This is not the case, because there is no way for blockly to know what the display is to be able to update it. By default fields will act like the colour field, displaying the last valid value, and having a value of the last valid value. If a developer wants to display invalid values (or put up warnings etc) they will need to override doValueInvalid_

RoboErikG replied:
We weren't very consistent before and I don't think there's a big difference between the two behaviors. I'm happy with this as is given that this makes it consistent and adds a way to override the default behavior.

BeksOmega commented on core/field_angle.js:
```diff
@@ -268,24 +272,15 @@ Blockly.FieldAngle.prototype.onMouseMove = function(e) {
+    Blockly.FieldTextInput.htmlInput_.value = angle;
+    this.setValue(angle);
+    // Always render the input angle.
+    this.text_ = String(angle);
```
Fixed

BeksOmega commented on core/field_variable.js:
```diff
@@ -192,30 +207,52 @@ Blockly.FieldVariable.prototype.getVariable = function() {
- *     variable.
+ * Gets the validation function for this field, or null if not set.
+ * Returns null if the variable is not set, because validators should not
+ * run on the initial setValue call.
```
Fixed

BeksOmega commented on core/field_variable.js:
```diff
@@ -192,30 +207,52 @@ Blockly.FieldVariable.prototype.getVariable = function() {
+  // this is not a possibility with variable fields.
+  if (this.variable_) {
+    return this.validator_;
+  }
```
Fixed

BeksOmega commented:
Ok, I resolved conversations from the last review that had been dealt with, just to keep things a bit more tidy.

I also tagged anything that needs an opinion with @rachel-fenichel  so that it'll be easier to for you to find (maybe, hopefully, idk how to github).

rachel-fenichel commented:
Sweet, this is getting really close.

Re: generator test failures
Travis shows this error:
https://travis-ci.org/google/blockly/jobs/536494731#L944
" WARN webdriver: Request failed due to Dismissed user prompt dialog: Error parsing XML: text.xml
InternalError: too much recursion"

So I opened the generator playground and checked only "text", then hit load.
That failed with an error that included `Blockly.Constants.Text.TEXT_CHARAT_EXTENSION`, so I knew it had to do with the `text_charAt` block:
![image](https://user-images.githubusercontent.com/13686399/58350151-f3bc5900-7e19-11e9-9dfb-05beb124c3e6.png)

For some reason it works if I load from XML on the playground with this XML:
```
<xml xmlns="http://www.w3.org/1999/xhtml">
<block type="text_charAt">
        <value name="VALUE">
          <block type="variables_get">
            <field name="VAR">text</field>
          </block>
        </value>
      </block>
</xml>
```
but fails if I drag in from the flyout.

The error looks like this:
```
connection.js:39 Uncaught RangeError: Maximum call stack size exceeded
    at Blockly.RenderedConnection.Blockly.Connection (connection.js:39)
    at new Blockly.RenderedConnection (rendered_connection.js:43)
    at Blockly.BlockSvg.makeConnection_ (block_svg.js:1418)
    at Blockly.BlockSvg.Blockly.Block.appendInput_ (block.js:1681)
    at Blockly.BlockSvg.appendInput_ (block_svg.js:1370)
    at Blockly.BlockSvg.Blockly.Block.appendValueInput (block.js:1379)
    at Blockly.BlockSvg.updateAt_ (text.js:840)
    at Blockly.FieldDropdown.validator_ (text.js:866)
    at Blockly.FieldDropdown.Blockly.Field.setValue (field.js:687)
    at Blockly.BlockSvg.Blockly.Block.setFieldValue (block.js:1129)
```


Anyway, that's not a solution, but should hopefully point you in the right direction for debugging.

Cheers,
Rachel

BeksOmega commented:
So the issues with generator tests were validators calling setValue when they shouldn't.

With regard to the text char block, the comment was actually incorrect, the dropdown never gets removed. You can see this if you external-ize the inputs.

![GenTests_TextAt](https://user-images.githubusercontent.com/25440652/58355818-71886080-7e2a-11e9-9690-7eedc8475c81.gif)

So that could safely be removed.

And with regard to the list_split block.
A) the setValue call wasn't included until #2326 
B) updateType_ only gets called when the dropdown is being validated (so the value will get updated anyway) or when domToMutation is called, and in that case the value should be set through the XML parsing.

![GenTests_ListSplit](https://user-images.githubusercontent.com/25440652/58356149-5c600180-7e2b-11e9-8f85-425bd03b3ab0.gif)

(same behavior as on develop - the shadow block was included to show #2326 functioning)

so that could be safely removed.

rachel-fenichel commented:
Makes sense.  Ready to merge?

BeksOmega commented:
All review comments have been dealt with! So It should be!

rachel-fenichel commented:
Here we go!  Thanks again for all your careful code and followups.

![](https://i.imgur.com/L03LulG.gif)
